src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java

```diff
@@ -534,9 +534,12 @@ synchronized boolean switchToAsyncMode() {
                 @Override public void onSuccess(CpsThreadGroup result) {
                     try {
                         // TODO keep track of whether the program was saved anyway after saveState was called but before now, and do not bother resaving it in that case
```

Comment:
BTW how does durability interact with `saveState`? Important.

Reply:
@jglick `CpsStepContext#saveState()` follows the same rules as overall Program durability - if your `DurabilityHint` has `isPersistWithEveryStep()` (i.e. `MAX_SURVIVABILITY` or `SURVIVABLE_NONATOMIC`) it saves normally.  

Otherwise (for `PERFORMANCE_OPTIMIZED`) it doesn't -- your program is not providing guaranteed persistence in the face of catastrophic system failures.  

Of course with a clean shutdown the pipeline will persist via the `@Terminator` as part of shutdown, so that's fine.